<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>StabMap: Stabilised mosaic single cell data integration using unshared features • StabMap</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="StabMap: Stabilised mosaic single cell data integration using unshared features">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-dark" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">StabMap</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/stabMap_PBMC_Multiome.html">StabMap: Stabilised mosaic single cell data integration using unshared features</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/sydneybiox/StabMap/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>StabMap: Stabilised mosaic single cell data integration using unshared features</h1>
                        <h4 data-toc-skip class="author">Shila
Ghazanfar</h4>
                        <h4 data-toc-skip class="author">Nick
Robertson</h4>
                        <h4 data-toc-skip class="author">Aiden Jin</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/sydneybiox/StabMap/blob/main/vignettes/stabMap_PBMC_Multiome.Rmd" class="external-link"><code>vignettes/stabMap_PBMC_Multiome.Rmd</code></a></small>
      <div class="d-none name"><code>stabMap_PBMC_Multiome.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://sydneybiox.github.io/StabMap">StabMap</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org" class="external-link">magrittr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellMultiModal</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">gridExtra</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2024</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>StabMap is a technique for performing mosaic single cell data
integration. Mosaic data integration presents the challenge of
integration of data where only some features or cells are shared across
datasets. For example, these challenges arise when integrating
single-cell datasets that measure different molecular profiles, such as
chromatin accessibility or RNA expression assays. Integrative analysis
of such data may provide a more in-depth profile of each cell,
facilitating downstream analysis. To read more about StabMap please see
our <a href="https://www.nature.com/articles/s41587-023-01766-z" class="external-link">paper
on Nature Biotechnology</a>.</p>
<div class="section level3">
<h3 id="vignette-goals">Vignette Goals<a class="anchor" aria-label="anchor" href="#vignette-goals"></a>
</h3>
<p>In this vignette we will elaborate on how mosaic single cell data
integration is implemented in the <code>StabMap</code> package. We
address a few key goals:</p>
<ul>
<li><p>Mosaic Data integration for 2 datasets</p></li>
<li><p>Demonstrating cell imputation following integration</p></li>
<li><p>Indirect mosaic data integration for 3 datasets, including 2
non-overlapping datasets</p></li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="load-data">Load data<a class="anchor" aria-label="anchor" href="#load-data"></a>
</h2>
<p>In this tutorial we will work with a multi-assay single cell dataset,
consisting of ATAC and gene expression data for 10,032 cells.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mae</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellMultiModal/man/scMultiome.html" class="external-link">scMultiome</a></span><span class="op">(</span></span>
<span>  <span class="st">"pbmc_10x"</span>,</span>
<span>  mode <span class="op">=</span> <span class="st">"*"</span>, dry.run <span class="op">=</span> <span class="cn">FALSE</span>, format <span class="op">=</span> <span class="st">"MTX"</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Perform some exploration of this data.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mae</span></span></code></pre></div>
<pre><code><span><span class="co">## A MultiAssayExperiment object of 2 listed</span></span>
<span><span class="co">##  experiments with user-defined names and respective classes.</span></span>
<span><span class="co">##  Containing an ExperimentList class object of length 2:</span></span>
<span><span class="co">##  [1] atac: SingleCellExperiment with 108344 rows and 10032 columns</span></span>
<span><span class="co">##  [2] rna: SingleCellExperiment with 36549 rows and 10032 columns</span></span>
<span><span class="co">## Functionality:</span></span>
<span><span class="co">##  experiments() - obtain the ExperimentList instance</span></span>
<span><span class="co">##  colData() - the primary/phenotype DataFrame</span></span>
<span><span class="co">##  sampleMap() - the sample coordination DataFrame</span></span>
<span><span class="co">##  `$`, `[`, `[[` - extract colData columns, subset, or experiment</span></span>
<span><span class="co">##  *Format() - convert into a long or wide DataFrame</span></span>
<span><span class="co">##  assays() - convert ExperimentList to a SimpleList of matrices</span></span>
<span><span class="co">##  exportClass() - save data to flat files</span></span></code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="http://waldronlab.io/MultiAssayExperiment/reference/upsetSamples.html" class="external-link">upsetSamples</a></span><span class="op">(</span><span class="va">mae</span><span class="op">)</span></span></code></pre></div>
<p><img src="stabMap_PBMC_Multiome_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">mae</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 6 rows and 6 columns</span></span>
<span><span class="co">##                  nCount_RNA nFeature_RNA nCount_ATAC nFeature_ATAC</span></span>
<span><span class="co">##                   &lt;integer&gt;    &lt;integer&gt;   &lt;integer&gt;     &lt;integer&gt;</span></span>
<span><span class="co">## AAACAGCCAAGGAATC       8380         3308       55582         13878</span></span>
<span><span class="co">## AAACAGCCAATCCCTT       3771         1896       20495          7253</span></span>
<span><span class="co">## AAACAGCCAATGCGCT       6876         2904       16674          6528</span></span>
<span><span class="co">## AAACAGCCAGTAGGTG       7614         3061       39454         11633</span></span>
<span><span class="co">## AAACAGCCAGTTTACG       3633         1691       20523          7245</span></span>
<span><span class="co">## AAACAGCCATCCAGGT       7782         3028       22412          8602</span></span>
<span><span class="co">##                                celltype broad_celltype</span></span>
<span><span class="co">##                             &lt;character&gt;    &lt;character&gt;</span></span>
<span><span class="co">## AAACAGCCAAGGAATC      naive CD4 T cells       Lymphoid</span></span>
<span><span class="co">## AAACAGCCAATCCCTT     memory CD4 T cells       Lymphoid</span></span>
<span><span class="co">## AAACAGCCAATGCGCT      naive CD4 T cells       Lymphoid</span></span>
<span><span class="co">## AAACAGCCAGTAGGTG      naive CD4 T cells       Lymphoid</span></span>
<span><span class="co">## AAACAGCCAGTTTACG     memory CD4 T cells       Lymphoid</span></span>
<span><span class="co">## AAACAGCCATCCAGGT non-classical monocy..        Myeloid</span></span></code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="fu"><a href="http://waldronlab.io/MultiAssayExperiment/reference/MultiAssayExperiment-methods.html" class="external-link">experiments</a></span><span class="op">(</span><span class="va">mae</span><span class="op">)</span><span class="op">[[</span><span class="st">"rna"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 36549 10032</span></span></code></pre>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="http://waldronlab.io/MultiAssayExperiment/reference/MultiAssayExperiment-methods.html" class="external-link">experiments</a></span><span class="op">(</span><span class="va">mae</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "atac" "rna"</span></span></code></pre>
<p>Keep the first 2,000 cells only. Normalise and select variable
features for the RNA modality.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce.rna</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://waldronlab.io/MultiAssayExperiment/reference/MultiAssayExperiment-methods.html" class="external-link">experiments</a></span><span class="op">(</span><span class="va">mae</span><span class="op">)</span><span class="op">[[</span><span class="st">"rna"</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Normalisation</span></span>
<span><span class="va">sce.rna</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sce.rna</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Feature selection</span></span>
<span><span class="va">decomp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html" class="external-link">modelGeneVar</a></span><span class="op">(</span><span class="va">sce.rna</span><span class="op">)</span></span>
<span><span class="va">hvgs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">decomp</span><span class="op">)</span><span class="op">[</span><span class="va">decomp</span><span class="op">$</span><span class="va">mean</span> <span class="op">&gt;</span> <span class="fl">0.01</span> <span class="op">&amp;</span> <span class="va">decomp</span><span class="op">$</span><span class="va">p.value</span> <span class="op">&lt;=</span> <span class="fl">0.05</span><span class="op">]</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">hvgs</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 952</span></span></code></pre>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce.rna</span> <span class="op">&lt;-</span> <span class="va">sce.rna</span><span class="op">[</span><span class="va">hvgs</span>, <span class="op">]</span></span></code></pre></div>
<p>Keep the first 2,000 cells only. Normalise and select variable
features for the ATAC modality.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="fu"><a href="http://waldronlab.io/MultiAssayExperiment/reference/MultiAssayExperiment-methods.html" class="external-link">experiments</a></span><span class="op">(</span><span class="va">mae</span><span class="op">)</span><span class="op">[[</span><span class="st">"atac"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 108344  10032</span></span></code></pre>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce.atac</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://waldronlab.io/MultiAssayExperiment/reference/MultiAssayExperiment-methods.html" class="external-link">experiments</a></span><span class="op">(</span><span class="va">mae</span><span class="op">)</span><span class="op">[[</span><span class="st">"atac"</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Normalise</span></span>
<span><span class="va">sce.atac</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sce.atac</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Feature selection using highly variable peaks</span></span>
<span><span class="co"># And adding matching peaks to genes</span></span>
<span><span class="va">decomp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html" class="external-link">modelGeneVar</a></span><span class="op">(</span><span class="va">sce.atac</span><span class="op">)</span></span>
<span><span class="va">hvgs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">decomp</span><span class="op">)</span><span class="op">[</span><span class="va">decomp</span><span class="op">$</span><span class="va">mean</span> <span class="op">&gt;</span> <span class="fl">0.25</span> <span class="op">&amp;</span></span>
<span>  <span class="va">decomp</span><span class="op">$</span><span class="va">p.value</span> <span class="op">&lt;=</span> <span class="fl">0.05</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">hvgs</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 788</span></span></code></pre>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce.atac</span> <span class="op">&lt;-</span> <span class="va">sce.atac</span><span class="op">[</span><span class="va">hvgs</span>, <span class="op">]</span></span></code></pre></div>
<p>Create a composite full data matrix by concatenating.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">logcounts_all</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">logcounts</a></span><span class="op">(</span><span class="va">sce.rna</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">logcounts</a></span><span class="op">(</span><span class="va">sce.atac</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">logcounts_all</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1]  1740 10032</span></span></code></pre>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">assayType</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">logcounts_all</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sce.rna</span><span class="op">)</span>,</span>
<span>  <span class="st">"rna"</span>, <span class="st">"atac"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">assayType</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## assayType</span></span>
<span><span class="co">## atac  rna </span></span>
<span><span class="co">##  788  952</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="mosaic-data-integration-with-stabmap">Mosaic data integration with StabMap<a class="anchor" aria-label="anchor" href="#mosaic-data-integration-with-stabmap"></a>
</h2>
<p>We will simulate a situation where half of the cells correspond to
the Multiome (RNA + ATAC features) modality, and half of the cells
correspond to the RNA modality. Our goal is to then integrate both
datasets by generating a joint embedding of the cells using all data,
and to impute the missing ATAC cell values from the RNA modality
cells.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dataType</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"RNA"</span>, <span class="st">"Multiome"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">logcounts_all</span><span class="op">)</span>,</span>
<span>    prob <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">0.5</span><span class="op">)</span>, replace <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">logcounts_all</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">dataType</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## dataType</span></span>
<span><span class="co">## Multiome      RNA </span></span>
<span><span class="co">##     5025     5007</span></span></code></pre>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">assay_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  RNA <span class="op">=</span> <span class="va">logcounts_all</span><span class="op">[</span><span class="va">assayType</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rna"</span><span class="op">)</span>, <span class="va">dataType</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"RNA"</span><span class="op">)</span><span class="op">]</span>,</span>
<span>  Multiome <span class="op">=</span> <span class="va">logcounts_all</span><span class="op">[</span></span>
<span>    <span class="va">assayType</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rna"</span>, <span class="st">"atac"</span><span class="op">)</span>, <span class="va">dataType</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Multiome"</span><span class="op">)</span></span>
<span>  <span class="op">]</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">assay_list</span>, <span class="va">dim</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $RNA</span></span>
<span><span class="co">## [1]  952 5007</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $Multiome</span></span>
<span><span class="co">## [1] 1740 5025</span></span></code></pre>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">assay_list</span>, <span class="va">class</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $RNA</span></span>
<span><span class="co">## [1] "dgCMatrix"</span></span>
<span><span class="co">## attr(,"package")</span></span>
<span><span class="co">## [1] "Matrix"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $Multiome</span></span>
<span><span class="co">## [1] "dgCMatrix"</span></span>
<span><span class="co">## attr(,"package")</span></span>
<span><span class="co">## [1] "Matrix"</span></span></code></pre>
<p>Examine the shared features between the two datasets using
<code><a href="../reference/mosaicDataUpSet.html">mosaicDataUpSet()</a></code>.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/mosaicDataUpSet.html">mosaicDataUpSet</a></span><span class="op">(</span><span class="va">assay_list</span>, plot <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="stabMap_PBMC_Multiome_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<p>From this we note that there are shared features between the RNA and
Multiome datasets, but there are many features that are observed only in
the Multiome dataset and not the RNA - as we had constructed.</p>
<p>We can understand the <code><a href="../reference/mosaicDataTopology.html">mosaicDataTopology()</a></code> of these
datasets, which generates an <code>igraph</code> object, which can be
inspected and plotted. The <code><a href="../reference/mosaicDataTopology.html">mosaicDataTopology()</a></code> is a
weighted network where nodes represent each dataset, and edges connect
nodes with at least one overlapping feature.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mdt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mosaicDataTopology.html">mosaicDataTopology</a></span><span class="op">(</span><span class="va">assay_list</span><span class="op">)</span></span>
<span><span class="va">mdt</span></span></code></pre></div>
<pre><code><span><span class="co">## IGRAPH 45affb8 UN-- 2 1 -- </span></span>
<span><span class="co">## + attr: name (v/c), frame.color (v/c), color (v/c), label.color (v/c),</span></span>
<span><span class="co">## | label.family (v/c)</span></span>
<span><span class="co">## + edge from 45affb8 (vertex names):</span></span>
<span><span class="co">## [1] RNA--Multiome</span></span></code></pre>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">mdt</span><span class="op">)</span></span></code></pre></div>
<p><img src="stabMap_PBMC_Multiome_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
<p>From this we note that the datasets RNA and Multiome share at least
some features. StabMap requires that the mosaic data topology network be
connected, that is, that there should be a path between every pair of
nodes in the network.</p>
<p>We now aim to integrate the data from the RNA and Multiome modality
by generating a common joint embedding for these data using
<code><a href="../reference/stabMap.html">stabMap()</a></code>. The <code><a href="../reference/stabMap.html">stabMap()</a></code> integration approach
aims to stabilize integration of single-cell data by exploting the
non-overlapping features, so that cells with similar biological profiles
will cluster. Stabilisation using non-overlapping features may be
important when there are limited overlapping features or when the
informative features are unknown.</p>
<details><summary><strong>What is <code>stabMap</code> doing?</strong>
</summary><p><code>stabMap</code> generates a joint embedding using 3 steps:</p>
<ul>
<li><p>Identify the <code><a href="../reference/mosaicDataTopology.html">mosaicDataTopology()</a></code></p></li>
<li><p>Embed the reference dataset into a lower dimensional
space</p></li>
<li><p>Project cells from non-reference datasets onto the reference
dataset embedding by using a model to traverse shortest paths in the
<code><a href="../reference/mosaicDataTopology.html">mosaicDataTopology()</a></code></p></li>
</ul></details><p>Since the Multiome data contains all features, we treat this as the
reference dataset. Since we already examined the mosaic data topology,
we set <code>plot = FALSE</code>.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">stab</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/stabMap.html">stabMap</a></span><span class="op">(</span><span class="va">assay_list</span>,</span>
<span>  reference_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Multiome"</span><span class="op">)</span>,</span>
<span>  plot <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## treating "Multiome" as reference</span></span></code></pre>
<pre><code><span><span class="co">## generating embedding for path with reference "Multiome": "Multiome"</span></span></code></pre>
<pre><code><span><span class="co">## generating embedding for path with reference "Multiome": "RNA" -&gt; "Multiome"</span></span></code></pre>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">stab</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 10032    50</span></span></code></pre>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">stab</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##                  Multiome_PC1 Multiome_PC2 Multiome_PC3 Multiome_PC4</span></span>
<span><span class="co">## AAACAGCCAATCCCTT    12.885344    -3.075968    -1.723863   -0.3561525</span></span>
<span><span class="co">## AAACAGCCAGTTTACG    11.314093    -2.344855     2.608507    1.2228681</span></span>
<span><span class="co">## AAACATGCAAGGTCCT    13.821325    -3.100703     4.755135   -0.6836924</span></span>
<span><span class="co">## AAACATGCACCGGCTA     6.287519    -2.080285   -24.802926   -0.6373922</span></span>
<span><span class="co">## AAACATGCAGCAAGTG    12.500354    -3.058831     5.358400   -2.6757611</span></span>
<span><span class="co">##                  Multiome_PC5</span></span>
<span><span class="co">## AAACAGCCAATCCCTT   -4.6468061</span></span>
<span><span class="co">## AAACAGCCAGTTTACG   -8.5576292</span></span>
<span><span class="co">## AAACATGCAAGGTCCT    6.0538837</span></span>
<span><span class="co">## AAACATGCACCGGCTA    7.1583625</span></span>
<span><span class="co">## AAACATGCAGCAAGTG   -0.1806992</span></span></code></pre>
<p>We can reduce the dimension further using non-linear approaches such
as UMAP.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">stab_umap</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">calculateUMAP</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">stab</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">stab_umap</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 10032     2</span></span></code></pre>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">stab_umap</span>, pch <span class="op">=</span> <span class="fl">16</span>, cex <span class="op">=</span> <span class="fl">0.3</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">dataType</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">stab</span><span class="op">)</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="stabMap_PBMC_Multiome_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
<p>Here we see that the RNA and Multiome cells are fairly
well-mixed.</p>
</div>
<div class="section level2">
<h2 id="data-imputation-after-stabmap">Data imputation after StabMap<a class="anchor" aria-label="anchor" href="#data-imputation-after-stabmap"></a>
</h2>
<p>Given the joint embedding, we can predict the missing ATAC cell
values using <code><a href="../reference/imputeEmbedding.html">imputeEmbedding()</a></code>. We use
<code><a href="../reference/imputeEmbedding.html">imputeEmbedding()</a></code> for demonstration purposes as for our
data both modalities have sufficient sample sizes (cells) and thus
cellular imputation isn’t needed.</p>
<p>To <code><a href="../reference/imputeEmbedding.html">imputeEmbedding()</a></code> we provide the data list, and the
joint embedding as output from <code><a href="../reference/stabMap.html">stabMap()</a></code>. We set the
Multiome cells as reference and the RNA cells as query. This is useful
for downstream visualisation or further interpretation.</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">imp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/imputeEmbedding.html">imputeEmbedding</a></span><span class="op">(</span></span>
<span>  <span class="va">assay_list</span>,</span>
<span>  <span class="va">stab</span>,</span>
<span>  reference <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">assay_list</span><span class="op">[[</span><span class="st">"Multiome"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>,</span>
<span>  query <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">assay_list</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">imp</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "list"</span></span></code></pre>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">imp</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Multiome"</span></span></code></pre>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">imp</span>, <span class="va">dim</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $Multiome</span></span>
<span><span class="co">## [1] 1740 5007</span></span></code></pre>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">assay_list</span>, <span class="va">dim</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $RNA</span></span>
<span><span class="co">## [1]  952 5007</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $Multiome</span></span>
<span><span class="co">## [1] 1740 5025</span></span></code></pre>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">imp</span><span class="op">[[</span><span class="st">"Multiome"</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## 5 x 5 sparse Matrix of class "dgCMatrix"</span></span>
<span><span class="co">##        AAACAGCCAAGGAATC AAACAGCCAATGCGCT AAACAGCCAGTAGGTG AAACAGCCATCCAGGT</span></span>
<span><span class="co">## CA6            1.299581         1.338925         1.075695                .</span></span>
<span><span class="co">## CNR2           .                .                .                       .</span></span>
<span><span class="co">## IFNLR1         .                .                .                       .</span></span>
<span><span class="co">## RCAN3          1.414502         1.553737         1.656583                .</span></span>
<span><span class="co">## ZNF683         .                .                .                       .</span></span>
<span><span class="co">##        AAACATGCACTTGTTC</span></span>
<span><span class="co">## CA6                   .</span></span>
<span><span class="co">## CNR2                  .</span></span>
<span><span class="co">## IFNLR1                .</span></span>
<span><span class="co">## RCAN3                 .</span></span>
<span><span class="co">## ZNF683                .</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="annotating-query-datasets-using-the-stabmap-embedding">Annotating Query Datasets using the StabMap embedding<a class="anchor" aria-label="anchor" href="#annotating-query-datasets-using-the-stabmap-embedding"></a>
</h2>
<p>We can also leverage this joint embedding to annotate the query data.
We will use a k-nearest neighbors (KNN) based algorithm to transfer cell
type labels from the reference to the query dataset. For our
demonstration we will treat the Multiome dataset as the reference and
the RNA dataset as the query.</p>
<p>The column data of the single cell experiments objects contained in
<code>mae</code> contain cell type annotations for each cell in the
<code>celltype</code> column. We first extract cell type annotations for
our reference dataset (Multiome).</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">annotation</span> <span class="op">&lt;-</span> <span class="st">"celltype"</span></span>
<span><span class="va">referenceLabels</span> <span class="op">&lt;-</span> <span class="fu">colData</span><span class="op">(</span></span>
<span>  <span class="fu"><a href="http://waldronlab.io/MultiAssayExperiment/reference/MultiAssayExperiment-methods.html" class="external-link">experiments</a></span><span class="op">(</span><span class="va">mae</span><span class="op">)</span><span class="op">[[</span><span class="st">"rna"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">assay_list</span><span class="op">$</span><span class="va">Multiome</span><span class="op">)</span>, <span class="va">annotation</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">referenceLabels</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">assay_list</span><span class="op">$</span><span class="va">Multiome</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">referenceLabels</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## referenceLabels</span></span>
<span><span class="co">##  CD56 (bright) NK cells     CD56 (dim) NK cells     classical monocytes </span></span>
<span><span class="co">##                     189                     217                     987 </span></span>
<span><span class="co">##    effector CD8 T cells  intermediate monocytes            MAIT T cells </span></span>
<span><span class="co">##                     205                     344                      49 </span></span>
<span><span class="co">##          memory B cells      memory CD4 T cells              myeloid DC </span></span>
<span><span class="co">##                     207                     792                     105 </span></span>
<span><span class="co">##           naive B cells       naive CD4 T cells       naive CD8 T cells </span></span>
<span><span class="co">##                     152                     745                     783 </span></span>
<span><span class="co">## non-classical monocytes         plasmacytoid DC </span></span>
<span><span class="co">##                     199                      51</span></span></code></pre>
<p>To classify query cells based on a reference dataset we can use the
<code><a href="../reference/classifyEmbedding.html">classifyEmbedding()</a></code> function. We provide the joint
embedding generated by <code><a href="../reference/stabMap.html">stabMap()</a></code> and cell type labels for
the reference dataset to the <code><a href="../reference/classifyEmbedding.html">classifyEmbedding()</a></code> function.
<code><a href="../reference/classifyEmbedding.html">classifyEmbedding()</a></code> returns a dataframe with predicted
labels in the <code>predicted_labels</code> column.</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">knn_out</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/classifyEmbedding.html">classifyEmbedding</a></span><span class="op">(</span></span>
<span>  <span class="va">stab</span>,</span>
<span>  <span class="va">referenceLabels</span>,</span>
<span><span class="op">)</span></span></code></pre></div>
<p>As we have simulated out datasets we have the true label annotations
for the RNA (query) cells. We can evaluate how well our predicted
annotations match the true annotations use a measure such as
accuracy.</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Extract query labels</span></span>
<span><span class="va">queryLabels</span> <span class="op">&lt;-</span> <span class="fu">colData</span><span class="op">(</span></span>
<span>  <span class="fu"><a href="http://waldronlab.io/MultiAssayExperiment/reference/MultiAssayExperiment-methods.html" class="external-link">experiments</a></span><span class="op">(</span><span class="va">mae</span><span class="op">)</span><span class="op">[[</span><span class="st">"rna"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">assay_list</span><span class="op">$</span><span class="va">RNA</span><span class="op">)</span>, <span class="va">annotation</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">queryLabels</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">assay_list</span><span class="op">$</span><span class="va">RNA</span><span class="op">)</span></span>
<span></span>
<span><span class="va">acc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">queryLabels</span> <span class="op">==</span> <span class="va">knn_out</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">queryLabels</span><span class="op">)</span>, <span class="st">"predicted_labels"</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">acc</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.9203116</span></span></code></pre>
<p>Since both the reference and query cells are embedded in the same low
dimensional space we can also visualise their cells together. Here we
present a UMAP visualisation colour coded by their cell types.</p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Extract reference and query cells from UMAP embedding</span></span>
<span><span class="va">stab_umap_ref</span> <span class="op">&lt;-</span> <span class="va">stab_umap</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">assay_list</span><span class="op">$</span><span class="va">Multiome</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="va">stab_umap_query</span> <span class="op">&lt;-</span> <span class="va">stab_umap</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">assay_list</span><span class="op">$</span><span class="va">RNA</span><span class="op">)</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># Create UMAP for reference cells</span></span>
<span><span class="va">df_umap_ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="va">stab_umap_ref</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>,</span>
<span>  y <span class="op">=</span> <span class="va">stab_umap_ref</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span>,</span>
<span>  cell_type <span class="op">=</span> <span class="va">referenceLabels</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">stab_umap_ref</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">p_ref</span> <span class="op">&lt;-</span> <span class="va">df_umap_ref</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, colour <span class="op">=</span> <span class="va">cell_type</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Reference cell type annotation"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create UMAP for query cells</span></span>
<span><span class="va">df_umap_query</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="va">stab_umap_query</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>,</span>
<span>  y <span class="op">=</span> <span class="va">stab_umap_query</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span>,</span>
<span>  cell_type <span class="op">=</span> <span class="va">queryLabels</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">stab_umap_query</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">p_query</span> <span class="op">&lt;-</span> <span class="va">df_umap_query</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, colour <span class="op">=</span> <span class="va">cell_type</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Query predicted cell types"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gridExtra/man/arrangeGrob.html" class="external-link">grid.arrange</a></span><span class="op">(</span><span class="va">p_ref</span>, <span class="va">p_query</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="stabMap_PBMC_Multiome_files/figure-html/unnamed-chunk-17-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="indirect-mosaic-data-integration-with-stabmap">Indirect mosaic data integration with StabMap<a class="anchor" aria-label="anchor" href="#indirect-mosaic-data-integration-with-stabmap"></a>
</h2>
<p>StabMap is a flexible framework for mosaic data integration, and can
still integrate data even when there are pairs of datasets that share no
features at all. So long as there is a path connecting the datasets
along the mosaic data topology (and the underlying assumption that the
shared features along these paths contain information), then we can
extract meaningful joint embeddings. To demonstrate this, we will
simulate three data sources.</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dataTypeIndirect</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"RNA"</span>, <span class="st">"Multiome"</span>, <span class="st">"ATAC"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">logcounts_all</span><span class="op">)</span>,</span>
<span>    prob <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.3</span>, <span class="fl">0.3</span>, <span class="fl">0.3</span><span class="op">)</span>, replace <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">logcounts_all</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">dataTypeIndirect</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## dataTypeIndirect</span></span>
<span><span class="co">##     ATAC Multiome      RNA </span></span>
<span><span class="co">##     3446     3342     3244</span></span></code></pre>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">assay_list_indirect</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  RNA <span class="op">=</span> <span class="va">logcounts_all</span><span class="op">[</span><span class="va">assayType</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rna"</span><span class="op">)</span>, <span class="va">dataTypeIndirect</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"RNA"</span><span class="op">)</span><span class="op">]</span>,</span>
<span>  Multiome <span class="op">=</span> <span class="va">logcounts_all</span><span class="op">[</span></span>
<span>    <span class="va">assayType</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rna"</span>, <span class="st">"atac"</span><span class="op">)</span>, <span class="va">dataTypeIndirect</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Multiome"</span><span class="op">)</span></span>
<span>  <span class="op">]</span>,</span>
<span>  ATAC <span class="op">=</span> <span class="va">logcounts_all</span><span class="op">[</span></span>
<span>    <span class="va">assayType</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"atac"</span><span class="op">)</span>, <span class="va">dataTypeIndirect</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ATAC"</span><span class="op">)</span></span>
<span>  <span class="op">]</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">assay_list_indirect</span>, <span class="va">dim</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $RNA</span></span>
<span><span class="co">## [1]  952 3244</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $Multiome</span></span>
<span><span class="co">## [1] 1740 3342</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $ATAC</span></span>
<span><span class="co">## [1]  788 3446</span></span></code></pre>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">assay_list_indirect</span>, <span class="va">class</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $RNA</span></span>
<span><span class="co">## [1] "dgCMatrix"</span></span>
<span><span class="co">## attr(,"package")</span></span>
<span><span class="co">## [1] "Matrix"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $Multiome</span></span>
<span><span class="co">## [1] "dgCMatrix"</span></span>
<span><span class="co">## attr(,"package")</span></span>
<span><span class="co">## [1] "Matrix"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $ATAC</span></span>
<span><span class="co">## [1] "dgCMatrix"</span></span>
<span><span class="co">## attr(,"package")</span></span>
<span><span class="co">## [1] "Matrix"</span></span></code></pre>
<p>Using <code><a href="../reference/mosaicDataUpSet.html">mosaicDataUpSet()</a></code>, we note that there are no
shared features between the ATAC and RNA datasets. For their integration
we might be able to match features by extracting genomic positions and
making the “central dogma assumption”, that is, that the peaks
associated with a genomic position overlapping a gene should correspond
to positive gene expression for that gene. However, using
<code><a href="../reference/stabMap.html">stabMap()</a></code> we need not make this assumption for the data
integration to be performed.</p>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/mosaicDataUpSet.html">mosaicDataUpSet</a></span><span class="op">(</span><span class="va">assay_list_indirect</span>, plot <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="stabMap_PBMC_Multiome_files/figure-html/unnamed-chunk-19-1.png" width="700"></p>
<p>We can understand the <code><a href="../reference/mosaicDataTopology.html">mosaicDataTopology()</a></code> of these
datasets, which generates an <code>igraph</code> object, which can be
inspected and plotted.</p>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mdt_indirect</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mosaicDataTopology.html">mosaicDataTopology</a></span><span class="op">(</span><span class="va">assay_list_indirect</span><span class="op">)</span></span>
<span><span class="va">mdt_indirect</span></span></code></pre></div>
<pre><code><span><span class="co">## IGRAPH aeeb398 UN-- 3 2 -- </span></span>
<span><span class="co">## + attr: name (v/c), frame.color (v/c), color (v/c), label.color (v/c),</span></span>
<span><span class="co">## | label.family (v/c)</span></span>
<span><span class="co">## + edges from aeeb398 (vertex names):</span></span>
<span><span class="co">## [1] RNA     --Multiome Multiome--ATAC</span></span></code></pre>
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">mdt_indirect</span><span class="op">)</span></span></code></pre></div>
<p><img src="stabMap_PBMC_Multiome_files/figure-html/unnamed-chunk-20-1.png" width="700"></p>
<p>StabMap only requires that the mosaic data topology network be
connected, that is, that there should be a path between every pair of
nodes in the network. While ATAC and RNA have no overlapping features,
since there is a path between RNA and ATAC (via Multiome), we can
proceed.</p>
<p>We now generate a common joint embedding for these data using
<code><a href="../reference/stabMap.html">stabMap()</a></code>. Since the Multiome data contains all features,
we again treat this as the reference dataset. Since we already examined
the mosaic data topology, we set <code>plot = FALSE</code>.</p>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">stab_indirect</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/stabMap.html">stabMap</a></span><span class="op">(</span><span class="va">assay_list_indirect</span>,</span>
<span>  reference_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Multiome"</span><span class="op">)</span>,</span>
<span>  plot <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## treating "Multiome" as reference</span></span></code></pre>
<pre><code><span><span class="co">## generating embedding for path with reference "Multiome": "Multiome"</span></span></code></pre>
<pre><code><span><span class="co">## generating embedding for path with reference "Multiome": "RNA" -&gt; "Multiome"</span></span></code></pre>
<pre><code><span><span class="co">## generating embedding for path with reference "Multiome": "ATAC" -&gt; "Multiome"</span></span></code></pre>
<div class="sourceCode" id="cb77"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">stab_indirect</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 10032    50</span></span></code></pre>
<div class="sourceCode" id="cb79"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">stab_indirect</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##                  Multiome_PC1 Multiome_PC2 Multiome_PC3 Multiome_PC4</span></span>
<span><span class="co">## AAACAGCCAATCCCTT    12.742178    3.0909138     1.757548    0.2695901</span></span>
<span><span class="co">## AAACAGCCAATGCGCT    12.708202    1.5955101    -6.170703    0.6596048</span></span>
<span><span class="co">## AAACATGCAAGGTCCT    13.646213    2.9647341    -4.846989    0.5665123</span></span>
<span><span class="co">## AAACATGCACCGGCTA     6.143255    2.1167652    24.572169    0.4845784</span></span>
<span><span class="co">## AAACATGCACTTGTTC     9.236392    0.8773643    -3.033358    0.2705357</span></span>
<span><span class="co">##                  Multiome_PC5</span></span>
<span><span class="co">## AAACAGCCAATCCCTT    4.4791329</span></span>
<span><span class="co">## AAACAGCCAATGCGCT   -2.3739650</span></span>
<span><span class="co">## AAACATGCAAGGTCCT   -6.5805017</span></span>
<span><span class="co">## AAACATGCACCGGCTA   -7.6081000</span></span>
<span><span class="co">## AAACATGCACTTGTTC    0.2939029</span></span></code></pre>
<p>We can reduce the dimension further using non-linear approaches such
as UMAP.</p>
<div class="sourceCode" id="cb81"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">stab_indirect_umap</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">calculateUMAP</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">stab_indirect</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">stab_indirect_umap</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 10032     2</span></span></code></pre>
<div class="sourceCode" id="cb83"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">stab_indirect_umap</span>,</span>
<span>  pch <span class="op">=</span> <span class="fl">16</span>, cex <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>  col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">dataTypeIndirect</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">stab_indirect</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="stabMap_PBMC_Multiome_files/figure-html/unnamed-chunk-22-1.png" width="700"></p>
<p>Here we see that the RNA, ATAC and Multiome cells are fairly
well-mixed.</p>
<p>Colouring the cells by their original cell type, we can also see that
the mosaic data integration is meaningful.</p>
<div class="sourceCode" id="cb84"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cellType</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="va">mae</span><span class="op">$</span><span class="va">celltype</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">mae</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">stab_indirect_umap</span>,</span>
<span>  pch <span class="op">=</span> <span class="fl">16</span>, cex <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>  col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">cellType</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">stab_indirect</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="stabMap_PBMC_Multiome_files/figure-html/unnamed-chunk-23-1.png" width="700"></p>
<details><summary><strong>Session Info</strong>
</summary><div class="sourceCode" id="cb85"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.4 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] HDF5Array_1.32.0            rhdf5_2.48.0               </span></span>
<span><span class="co">##  [3] DelayedArray_0.30.1         SparseArray_1.4.8          </span></span>
<span><span class="co">##  [5] S4Arrays_1.4.1              abind_1.4-5                </span></span>
<span><span class="co">##  [7] Matrix_1.7-0                gridExtra_2.3              </span></span>
<span><span class="co">##  [9] SingleCellMultiModal_1.16.0 MultiAssayExperiment_1.30.3</span></span>
<span><span class="co">## [11] scran_1.32.0                scater_1.32.1              </span></span>
<span><span class="co">## [13] ggplot2_3.5.1               scuttle_1.14.0             </span></span>
<span><span class="co">## [15] SingleCellExperiment_1.26.0 SummarizedExperiment_1.34.0</span></span>
<span><span class="co">## [17] Biobase_2.64.0              GenomicRanges_1.56.1       </span></span>
<span><span class="co">## [19] GenomeInfoDb_1.40.1         IRanges_2.38.1             </span></span>
<span><span class="co">## [21] S4Vectors_0.42.1            BiocGenerics_0.50.0        </span></span>
<span><span class="co">## [23] MatrixGenerics_1.16.0       matrixStats_1.3.0          </span></span>
<span><span class="co">## [25] magrittr_2.0.3              StabMap_0.99.3             </span></span>
<span><span class="co">## [27] BiocStyle_2.32.1           </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] jsonlite_1.8.8            ggbeeswarm_0.7.2         </span></span>
<span><span class="co">##   [3] magick_2.8.4              farver_2.1.2             </span></span>
<span><span class="co">##   [5] rmarkdown_2.27            fs_1.6.4                 </span></span>
<span><span class="co">##   [7] zlibbioc_1.50.0           ragg_1.3.2               </span></span>
<span><span class="co">##   [9] vctrs_0.6.5               memoise_2.0.1            </span></span>
<span><span class="co">##  [11] DelayedMatrixStats_1.26.0 htmltools_0.5.8.1        </span></span>
<span><span class="co">##  [13] BiocBaseUtils_1.6.0       AnnotationHub_3.12.0     </span></span>
<span><span class="co">##  [15] curl_5.2.1                BiocNeighbors_1.22.0     </span></span>
<span><span class="co">##  [17] Rhdf5lib_1.26.0           sass_0.4.9               </span></span>
<span><span class="co">##  [19] bslib_0.8.0               desc_1.4.3               </span></span>
<span><span class="co">##  [21] plyr_1.8.9                cachem_1.1.0             </span></span>
<span><span class="co">##  [23] igraph_2.0.3              mime_0.12                </span></span>
<span><span class="co">##  [25] lifecycle_1.0.4           pkgconfig_2.0.3          </span></span>
<span><span class="co">##  [27] rsvd_1.0.5                R6_2.5.1                 </span></span>
<span><span class="co">##  [29] fastmap_1.2.0             GenomeInfoDbData_1.2.12  </span></span>
<span><span class="co">##  [31] digest_0.6.36             colorspace_2.1-1         </span></span>
<span><span class="co">##  [33] AnnotationDbi_1.66.0      dqrng_0.4.1              </span></span>
<span><span class="co">##  [35] irlba_2.3.5.1             ExperimentHub_2.12.0     </span></span>
<span><span class="co">##  [37] textshaping_0.4.0         RSQLite_2.3.7            </span></span>
<span><span class="co">##  [39] beachmat_2.20.0           labeling_0.4.3           </span></span>
<span><span class="co">##  [41] filelock_1.0.3            fansi_1.0.6              </span></span>
<span><span class="co">##  [43] httr_1.4.7                compiler_4.4.1           </span></span>
<span><span class="co">##  [45] bit64_4.0.5               withr_3.0.0              </span></span>
<span><span class="co">##  [47] BiocParallel_1.38.0       viridis_0.6.5            </span></span>
<span><span class="co">##  [49] DBI_1.2.3                 UpSetR_1.4.0             </span></span>
<span><span class="co">##  [51] highr_0.11                rappdirs_0.3.3           </span></span>
<span><span class="co">##  [53] rjson_0.2.21              bluster_1.14.0           </span></span>
<span><span class="co">##  [55] tools_4.4.1               vipor_0.4.7              </span></span>
<span><span class="co">##  [57] beeswarm_0.4.0            glue_1.7.0               </span></span>
<span><span class="co">##  [59] rhdf5filters_1.16.0       grid_4.4.1               </span></span>
<span><span class="co">##  [61] cluster_2.1.6             generics_0.1.3           </span></span>
<span><span class="co">##  [63] gtable_0.3.5              BiocSingular_1.20.0      </span></span>
<span><span class="co">##  [65] ScaledMatrix_1.12.0       metapod_1.12.0           </span></span>
<span><span class="co">##  [67] utf8_1.2.4                XVector_0.44.0           </span></span>
<span><span class="co">##  [69] RcppAnnoy_0.0.22          ggrepel_0.9.5            </span></span>
<span><span class="co">##  [71] BiocVersion_3.19.1        pillar_1.9.0             </span></span>
<span><span class="co">##  [73] limma_3.60.4              dplyr_1.1.4              </span></span>
<span><span class="co">##  [75] BiocFileCache_2.12.0      lattice_0.22-6           </span></span>
<span><span class="co">##  [77] bit_4.0.5                 tidyselect_1.2.1         </span></span>
<span><span class="co">##  [79] locfit_1.5-9.10           Biostrings_2.72.1        </span></span>
<span><span class="co">##  [81] knitr_1.48                bookdown_0.40            </span></span>
<span><span class="co">##  [83] edgeR_4.2.1               xfun_0.46                </span></span>
<span><span class="co">##  [85] statmod_1.5.0             UCSC.utils_1.0.0         </span></span>
<span><span class="co">##  [87] yaml_2.3.10               evaluate_0.24.0          </span></span>
<span><span class="co">##  [89] codetools_0.2-20          tibble_3.2.1             </span></span>
<span><span class="co">##  [91] BiocManager_1.30.23       cli_3.6.3                </span></span>
<span><span class="co">##  [93] uwot_0.2.2                systemfonts_1.1.0        </span></span>
<span><span class="co">##  [95] munsell_0.5.1             jquerylib_0.1.4          </span></span>
<span><span class="co">##  [97] Rcpp_1.0.13               dbplyr_2.5.0             </span></span>
<span><span class="co">##  [99] png_0.1-8                 parallel_4.4.1           </span></span>
<span><span class="co">## [101] pkgdown_2.1.0             blob_1.2.4               </span></span>
<span><span class="co">## [103] sparseMatrixStats_1.16.0  SpatialExperiment_1.14.0 </span></span>
<span><span class="co">## [105] slam_0.1-51               viridisLite_0.4.2        </span></span>
<span><span class="co">## [107] scales_1.3.0              purrr_1.0.2              </span></span>
<span><span class="co">## [109] crayon_1.5.3              rlang_1.1.4              </span></span>
<span><span class="co">## [111] KEGGREST_1.44.1</span></span></code></pre>
</details>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Shila Ghazanfar.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
